velox/connectors/hive/HiveWriteProtocol.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/HiveWriteProtocol.h"
+#include "velox/connectors/hive/HiveConnector.h"
+
+#include <boost/lexical_cast.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace facebook::velox::connector::hive {
+
+namespace {
+
+std::vector<std::shared_ptr<const HiveColumnHandle>> getPartitionColumns(
+    const std::shared_ptr<const HiveInsertTableHandle>& tableWriteHandle) {
+  std::vector<std::shared_ptr<const HiveColumnHandle>> partitionColumns;
+  for (const auto& column : tableWriteHandle->inputColumns()) {
+    const auto hiveColumn =
+        std::dynamic_pointer_cast<const HiveColumnHandle>(column);
+    if (hiveColumn->columnType() ==
+        HiveColumnHandle::ColumnType::kPartitionKey) {
+      partitionColumns.emplace_back(hiveColumn);
+    }
+  }
+  return partitionColumns;
+}
+
+std::string makeUuid() {
+  return boost::lexical_cast<std::string>(boost::uuids::random_generator()());
+}
+
+} // namespace
+
+std::shared_ptr<const WriterParameters>
+HiveNoCommitWriteProtocol::getWriterParameters(
+    const std::shared_ptr<const ConnectorInsertTableHandle>& tableWriteHandle,
+    const ConnectorQueryCtx* FOLLY_NONNULL connectorQueryCtx) const {
+  auto hiveTableWriteHandle =
+      std::dynamic_pointer_cast<const HiveInsertTableHandle>(tableWriteHandle);
+  if (!hiveTableWriteHandle) {
+    VELOX_UNSUPPORTED("This write protocol cannot used for non-Hive connector");
+  }
+
+  if (!getPartitionColumns(hiveTableWriteHandle).empty()) {
```

Comment:
If all you need to know is whether there are any partition columns, then it is more efficient to have a function that returns a boolean instead of allocating and populating a vector of partition columns, then checking if it is empty.

```
VELOX_USER_CHECK(hasPartitionColumns(hiveTableWriteHandle);
```

Reply:
Make sense to me! Revised.